compiler/rustc_borrowck/src/type_check/mod.rs

```diff
@@ -385,7 +385,14 @@ impl<'a, 'b, 'tcx> Visitor<'tcx> for TypeVerifier<'a, 'b, 'tcx> {
             } else if let Some(static_def_id) = constant.check_static_ptr(tcx) {
                 let unnormalized_ty = tcx.type_of(static_def_id).instantiate_identity();
```

Comment:
so the issue here is that `type_of(static_def_id)` is `ty::Error` here which does not have a `builtin_deref`? I would like to instead explicitly check for `unnormalized_ty.has_errors()` or whatever and bail in that case

Reply:
The problem is with `constant.const_.ty()`, not with `unnormalized_ty`

E.g. in the test below, `unnormalized_ty` is `u8`, which is perfectly fine. The type of the constant would be `*const u8`, but earlier logic from https://github.com/rust-lang/rust/pull/125558 turns that type into a type error because it is not an integer type. The error is inserted here https://github.com/Amanieu/rust/blob/eb726a50e65387c348ba56f2416fc3c0b649f95e/compiler/rustc_hir_analysis/src/collect/type_of.rs#L105

Reply:
so the issue is that `constant` has type `<type error>` but its value is `ConstValue::Scalar`?

I think we should try to instead also taint the value of the const to be an error :thinking:

Reply:
How does that work? 

Also in the future we'll likely want to be able to use pointers in this position https://github.com/rust-lang/rust/issues/128464 it's just not implemented right now and trying it should not ICE.

Which is to say: changes made now may be reverted when using pointers as asm const operands is properly implemented.

Reply:
At some point you create a constant whose type is `ty::Error` but which has an actual value. By also invalidating the value of the constant at this point `if let Some(static_def_id) = constant.check_static_ptr(tcx)` does not get triggered anymore.

Reply:
Sure, I meant more concretely, e.g. here https://github.com/Amanieu/rust/blob/eb726a50e65387c348ba56f2416fc3c0b649f95e/compiler/rustc_hir_analysis/src/collect/type_of.rs#L105 

how does one also invalidate the value? None of the other branches appear to do that

Reply:
well, I gave this a go. Th current version turns the basic block into a `TerminatorKind::Unreachable`. 

This feels extremely hacky though? more so than catching an unsatisfied assumption in the borrow checker. But maybe I just misunderstand what you mean by "invalidate"? Is there some way to poison a `Const` that I'm missing?

Reply:
yes, the idea would be to use a `ty::ConstKind::Error` as the const value

Reply:
cool, that looks a lot better!